src/async_impl/decoder.rs

```diff
@@ -365,10 +365,19 @@ impl HttpBody for Decoder {
             }
             #[cfg(feature = "gzip")]
             Inner::Gzip(ref mut decoder) => {
-                match futures_core::ready!(Pin::new(decoder).poll_next(cx)) {
+                match futures_core::ready!(Pin::new(&mut *decoder).poll_next(cx)) {
                     Some(Ok(bytes)) => Poll::Ready(Some(Ok(Frame::data(bytes.freeze())))),
                     Some(Err(err)) => Poll::Ready(Some(Err(crate::error::decode_io(err)))),
-                    None => Poll::Ready(None),
+                    None => { // poll inner connection until EOF after gzip stream is finished
+                        let gzip_decoder = decoder.get_mut();
+                        let stream_reader = gzip_decoder.get_mut();
+                        let peekable_io_stream = stream_reader.get_mut();
+                        match futures_core::ready!(Pin::new(peekable_io_stream).poll_next(cx)) {
+                            Some(Ok(bytes)) => Poll::Ready(Some(Ok(Frame::data(bytes)))),
```

Comment:
I think this is should likely be an error. If you said a response's `content-encoding: gzip`, it's assumed the entire content is encoded. If there's some dangling data that isn't, that seems bad.

Reply:
Could there be multiple consecutive chunks (`Transfer-Encoding: chunked`) in body?
And what about HTTP trailers?

Reply:
Yes, and most likely the compressed body came over multiple chunks. But `content-encoding` implies the entire content.

Trailers would come through as a different `Frame`, but the `IoStream` already tosses them. We could (in a separate PR, perhaps) reach into the `IoStream` and poll that instead, to support trailers.

Reply:
I mean, can it be like this?
```
|--------------original-content--------------------|
|-----gzip-content-1-----||-----gzip-content-2-----|
|-chunk-1-||---chunk-2---||----------chunk-3-------|
```

Reply:
If `Some(Ok(bytes))` should return error, then what kind of error?

Reply:
> I mean, can it be like this?

That would only work if using "multiple members", which is a separate idea.

> what kind of error?

Probably something like `error::decode("extra bytes after decompression")`. The exact message can be tweaked.